feat: add local HTTP API for transcription

Teyk0o opened the pull request:
# Related to : https://github.com/Kieirra/murmure/issues/2

### Description

This PR introduces a **local HTTP API** that allows external applications to transcribe audio files using Murmure **without the graphical interface**.
It enables Murmure to act as a **reusable, privacy-first Speech-to-Text engine**, easily integrated into other tools or workflows.

---

### Features

* **POST `/api/transcribe`** endpoint accepting `.wav` files via multipart form
* Experimental toggle in **Settings → System** to enable or disable the API
* **Configurable port** (default: `4800`) with validation (`1024–65535`)
* **Automatic language detection** via Parakeet
* **Dictionary corrections** applied to transcriptions
* **Audio resampling** to 16kHz if needed
* **JSON responses** for success and errors
* **CORS enabled** for cross-origin requests
* **Localhost-only binding** for privacy

---

### Backend Changes

* Added new `http_api.rs` module using the **Axum** web framework
* Made `transcribe_audio()` function public in `audio.rs`
* Added `api_enabled` and `api_port` fields to `AppSettings`
* Added `get/set` commands for API configuration
* Integrated HTTP server startup during app initialization

---

### Frontend Changes

* Added `useApiState` hook for managing API settings
* Added API toggle and port configuration in **System Settings**
* Displayed the API endpoint URL when enabled

---

### Dependencies Added

* `tokio` — async runtime
* `axum` — HTTP framework with multipart support
* `tower-http` — middleware and utilities
* `hyper` — HTTP protocol
* `uuid` — temporary file naming

---

### Testing

* Tested with **Postman** on Windows
* Successfully transcribes **French and English** audio
* Settings persist across app restarts
* No compilation warnings or errors
* Added tests/voices/ directory containing two test voices generated with ElevenLabs, reading a long text sample to validate performance and transcription accuracy.

<img width="1497" height="765" alt="image" src="https://github.com/user-attachments/assets/759112aa-38cc-4281-8162-e1f26722a6a0" />

Kieirra commented:
Wow, that was fast and from what I can see, really clean work.

I'll review and test it after work, then merge if everything looks good.

Kieirra commented on API_USAGE.md:
```diff
@@ -0,0 +1,101 @@
+# Murmure Local HTTP API
```
Nice doc

Kieirra commented on src-tauri/src/lib.rs:
```diff
@@ -75,6 +76,26 @@ pub fn run() {
             app.manage(TranscriptionSuspended::new(false));
 
+            // Start HTTP API if enabled
+            if s.api_enabled {
```
OK good! But it would be better if we could spawn the thread as a command when the user enables the option because if it only runs here, it will only work after a restart.

That might be fine, but in that case, we should make sure it’s clearly specified in the frontend.

Kieirra commented on src/features/settings/system/system.tsx:
```diff
@@ -110,6 +113,48 @@ export const System = () => {
+                                onCheckedChange={setApiEnabled}
+                            />
+                        </SettingsUI.Item>
+                        {apiEnabled && (
```
Nice

Kieirra commented on src/features/settings/system/system.tsx:
```diff
@@ -110,6 +113,48 @@ export const System = () => {
+                                <Typography.Paragraph>
+                                    Enable a local HTTP API for transcribing audio files
+                                    from other applications. Access it at
+                                    http://localhost:{apiPort}/api/transcribe
```
I’ll add a link to your documentation after merging this.

Kieirra commented on src-tauri/src/http_api.rs:
```diff
@@ -0,0 +1,151 @@
+    let router = Router::new()
+        .route("/api/transcribe", post(transcribe_handler))
+        .with_state(app.clone())
+        .layer(CorsLayer::permissive())
```
Hmm… this seems a bit risky, because in theory a random website you visit could make requests to Murmure, which isn’t a great idea in my opinion.

Does it still work if we remove this completely? Or maybe we could restrict it to only allow http://localhost and http://127.0.0.1?

Kieirra commented on src-tauri/src/lib.rs:
```diff
@@ -75,6 +76,26 @@ pub fn run() {
+                    match rt {
+                        Ok(runtime) => {
+                            if let Err(e) = runtime.block_on(http_api::start_http_api(app_handle, port)) {
+                                eprintln!("HTTP API error: {}", e);
```
It would be nice to notify the frontend with a dialog when the port is already in use, so the user understands what happened and can pick a new available port.

Teyk0o commented:
Thanks I will make it tomorrow!

Teyk0o commented:
### Update Summary – HTTP API PR

This update refines and finalizes the local HTTP API integration for Murmure, improving security, maintainability, and user experience based on reviewer feedback.

---

### New and Updated Commits

**chore: remove package-lock.json**
*Remove npm lockfile since pnpm is the package manager used.*

* Ensures consistent dependency management across environments.

**refactor: start HTTP API server on demand instead of at boot**
*Improves control over when the API starts.*

* Adds `start_http_api_server` command.
* Moves server startup from initialization to on-demand.
* API now starts instantly when toggled on in Settings.
* Resolves feedback about delayed API availability.

**refactor: make useApiState hook consistent with codebase patterns**
*Aligns frontend logic with established async state patterns.*

* Extracted `loadApiState` as async function.
* Replaced `.then()` with `async/await`.
* Added error handling and state rollback.
* Matches the behavior of other settings hooks.

**docs: improve API documentation with troubleshooting and limitations**
*Expands documentation for clarity and developer onboarding.*

* Added **Quick Start** section.
* Clarified automatic API start (no restart needed).
* Added **Troubleshooting** and **Limitations** sections.
* Documented sequential processing and language detection.

**security: remove CORS layer for better security**
*Hardens security by removing unnecessary CORS layer.*

* Since API binds to `127.0.0.1`, CORS is not needed.
* Removed permissive `CorsLayer` and unused dependencies (`tower`, `tower-http`).

**feat: notify user when HTTP API fails to start**
*Improves UX and transparency when startup errors occur.*

* Added dialog notifications for startup errors.
* Special handling for "address already in use" cases.
* Generic fallback message for other failures.
* Enhanced console logs with full error details.

**feat: add HTTP API server with auto-start and graceful shutdown**
*Implements the `/api/transcribe` endpoint with robust lifecycle handling.*

* Supports multipart audio uploads.
* Auto-starts when enabled in settings.
* Graceful shutdown with `tokio::select!` and oneshot channels.
* Auto-restarts when port changes.
* Added dialog notifications for startup errors (ports, etc...) using Tauri Dialog Plugin ([@tauri-apps/plugin-dialog](https://v2.tauri.app/fr/plugin/dialog/)).
* Enforces port validation (1024–65535).
* Cleans up temporary audio files post-transcription.

**chore: formatting and cleanup**
*Standardizes and cleans codebase post-implementation.*

* Unified line endings and formatting.
* Removed outdated API documentation.
* Added new API tests and docs directory.
* Applied consistent code style.

---

### Summary

This round of commits finalizes the **local HTTP API** feature with a stable architecture, clear documentation, and improved UX. The server now starts dynamically, handles errors gracefully, and maintains strong local-only security. The result: Murmure can now act as a **modular and privacy-first transcription engine**, ready for broader local integrations.
